### PR TITLE
add inplace attribute to op_proto_maker

### PR DESCRIPTION
**Pull request by dzhwinter**
- [x] add reused attribute to OpProto.
- [x] change the related operator to use inplace attribute.

### CODE REVIEW

File: paddle/fluid/operators/sum_op.cc [hunk 1]
```diff
@@ -115,7 +115,7 @@ class SumOpMaker : public framework::OpProtoAndCheckerMaker {
   void Make() override {
     AddInput("X", "(vector<Tensor>) The input tensors of sum operator.")
         .AsDuplicable();
-    AddOutput("Out", "(Tensor) The output tensor of sum operator.");
+    AddOutput("Out", "(Tensor) The output tensor of sum operator.").Reuse("X");
```
Comment (dzhwinter) [inline paddle/fluid/operators/sum_op.cc]: In this case, Out should reuse the first Tensor memory of X(vector<Tensor> ), but we didn't have the sub-variable name of X.

**Comment by tonyyang-svail** [comment]
The test only covers OpMaker, please add a runtime test for this feature.

**Comment by tonyyang-svail** [comment]
there are already a bunch of `in_place` field on [the python side](https://github.com/PaddlePaddle/Paddle/search?utf8=%E2%9C%93&q=in_place&type=). Not sure how to make them work together...

**Comment by dzhwinter** [comment]
Have discussed this issue with @tonyyang-svail , we will move the in-place attribute into the ssa-graph framework. This PR is add an inplace attribute for every Operator in Python-side.

File: paddle/fluid/operators/adam_op.cc [hunk 2]
```diff
@@ -89,9 +89,9 @@ class AdamOpMaker : public framework::OpProtoAndCheckerMaker {
     AddInput("Beta1Pow", "(Tensor) Input beta1 power accumulator");
     AddInput("Beta2Pow", "(Tensor) Input beta2 power accumulator");
 
-    AddOutput("ParamOut", "(Tensor) Output parameter");
-    AddOutput("Moment1Out", "(Tensor) Output first moment");
-    AddOutput("Moment2Out", "(Tensor) Output second moment");
+    AddOutput("ParamOut", "(Tensor) Output parameter").Reuse("Param");
+    AddOutput("Moment1Out", "(Tensor) Output first moment").Reuse("Moment1");
+    AddOutput("Moment2Out", "(Tensor) Output second moment").Reuse("Moment2");
```
Comment (JiayiFeng) [inline paddle/fluid/operators/adam_op.cc]: Is `Moment2Out` forced share memory with `Moment2`?

Comment (dzhwinter) [reply paddle/fluid/operators/adam_op.cc]: It's not *forced*, is a suggestion will be get in memory optimizer in Python.

**Pull request merged**